### ISSUE TITLE
Null conditional operators

### ISSUE THREAD

**Issue by Andrew-Hanlon** [dotnet/csharplang#1427]
Proposal to extend the null conditional syntax to apply to operators:

Instead of writing code like: 
```````````````````````````````
result = ( foo != null ? foo + "." : "" ) + "bar";

or

sum = val != null ? val + 5 : 0;
```````````````````````````````
Use syntax sugar to write:
``````````
result = (foo ?+ ".") + "bar";

or 

sum = (val ?+ 5) ?? 0;
``````````

This would:
- Eliminate possible duplicate evaluations,
- Allow the user to be explicit in intent without worrying about the implementation details of a type's operators,
- Reduce complexity and block usage for inline and lambda conditional operations .

**Comment by orthoxerox** [comment]
Similar to #431

**Comment by svick** [comment]
```c#
sum = val != null ? val + 5 : 0; // val+5 or 0 when null
sum = (val ?+ 5) ?? 0;
```

You can write this today as:

```c#
sum = (val + 5) ?? 0;
```

This works, because `+` is lifted for `int?`.

(You can't do that for your first case, because `null` behaves the same as the empty string with string concatenation.)

**Comment by Andrew-Hanlon** [comment]
@orthoxerox Insteresting - I wasn't thinking to that depth of general lifting operations, that seems more complex (though possibly this is too simplistic). 

@svick Good point, though it highlights my point about being explicit in intention, val could have been a different type (not a nullable int) with a different implementation.

**Comment by Joe4evr** [comment]
> val could have been a different type (not a nullable int) with a different implementation.

If it was, you'd get a compiler error because there is no `operator +` for `int` and a different type as its operands.

**Comment by Andrew-Hanlon** [comment]
@Joe4evr The operator precedence would be followed, so if 'val' *was* a custom `ComplexInt` with an operator like:
 ````````
public static int? operator +(ComplexInt c1, int c2)
{
	if(c1 == null)
	return -1;
	
	return c1.Value + c2;
}
`````````
It would return -1 instead of zero;

I understand that this is contrived and was a mistaken example on my part, but the point about explicit intentions is still accurate.

**Comment by svick** [comment]
@Andrew-Hanlon I don't think it makes sense to add a new operator that duplicates the behavior of an existing operator, just to make the intentions more explicit. Especially since those intentions are sill not completely explicit.

**Comment by Andrew-Hanlon** [comment]
@svick Appreciate your point of view - from my perspective, a 'null conditional *for* operators' would be as explicit as, and work consistently with, the 'null conditional operator' for member/method access. It fundamentally  maps `x ?- y` into something like `x?.Minus(y)`. Which to me is quite tidy, and cleanly avoids the null handling implementation details from operators.

**Comment by CyrusNajmabadi** [comment]
I'm a little confused:

> with, the 'null conditional operator' for member/method access. It fundamentally maps ```x ?- y``` into something like ```x?.Minus(y)```

Why do you need this?  The compiler already maps ```x - y``` in the case of anything nullable.  Why would you need another operator to do what is already being done? 

Thanks!

**Comment by Andrew-Hanlon** [comment]
@CyrusNajmabadi Any type can have operator overloads, not just value types or numerics, and there is no mandate that they handle nulls in a consistent way. For instance strings treat null as an empty string during concatenation. Obviously this isn't the highest priority, I just thought it added consistency. And for strings especially, would make 'conditional concatenation' much cleaner. It was a poor example chosen for the nullable integer - though it highlights at least that it wouldn't break the current lifting approach. Cheers.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Any type can have operator overloads, not just value types or numerics, and there is no mandate that they handle nulls in a consistent way.

Sure.... but if you want to handle null in a special way, then why not add an operator just for that?  

For example:

```c#
struct MyStruct
{
    private readonly int _value;

    public MyStruct(int value) => _value = value;

    public static MyStruct operator +(MyStruct opt, int i)
        => new MyStruct(opt._value + 1);

    public static MyStruct operator +(MyStruct? opt, int i)
        => opt == null ? new MyStruct(i) : new MyStruct(opt.Value._value + i);
}

// ...
        MyStruct? s = null;
        MyStruct t = s + 1;
```

The use of ```+```  in the last line will call your operator, where you can have whatever semantics you want.  

Here, i've defined ```+``` on a Nullable 'MyStruct' to mean that it's the same as if you were operating on hte 'zero' MyStruct'...

**Comment by Andrew-Hanlon** [comment]
@CyrusNajmabadi This 'null conditional operator' syntax would be meant to help with the consumption of _existing_ types - there are no operator extension methods... yet..

**Comment by CyrusNajmabadi** [comment]
@Andrew-Hanlon Sorry... i don't know what the above means.  can you clarify?  Are you basically saying: i have a null struct... and i want to handle operators on it *differently* than how the actual struct defined the operators?

**Comment by AustinBryan** [comment]
You can do this with extension methods

```csharp
result = (foo != null ? foo + "." : "" ) + "bar";          // "foo.bar" or "bar" when null
result = foo?.Add(".") + "bar";                            // "foo.bar" or "bar" when null

public static string Add(this string a, string b) => $"{(a == null ? "" : a + b)}";
```
